# Advent of Code Challenges - Day 7

## Load Libraries

In [1]:
import numpy as np
import pandas as pd

from pandas import Series,DataFrame

## Load Data

In [2]:
input_path = './input.txt'

with open(input_path) as f:
    lines = f.readlines()
f.close()

# Convert data to pandas Series
data = lines[0].split(",")
data = [int(x) for x in data]
series = Series(data)

## Define Functions

In [3]:
def calc_fuel_cost(n):
    if n == 0:
        fuel_cost = 0
    else:
        fuel_cost = Series([n - x for x in range(0,n,1)]).sum()
    return(fuel_cost)

## Puzzle 1 - Shellfish, to the rescue!
A large number of crabs joins your flight from Mobey Dick. Their objective, to blast a hole in the sea floor to open up a passage to a series of caves (Wow! That was lucky):
- The crabs must all be aligned to a single horizontal position
- This must be done in a way that minimizes fuel costs (crab submarines have limited fuel)

In [4]:
# in the example, the position selected to minimze fuel costs was 2.
# this also happens to be the median of the series. Let's see if the same holds true for the full dataset
sample_list = [16,1,2,0,4,2,7,1,2,14]
sample_series = Series(sample_list)
print(sample_series.median())

2.0


In [5]:
series_median = series.median()

df = DataFrame(series)
df.columns = ['start_pos']
df['fuel_cost'] = abs(series_median - df.start_pos)
total_fuel_cost = df['fuel_cost'].sum()
print(f"Answer: {int(total_fuel_cost)}")

Answer: 326132


## Puzzle 2 - Fuel costs keep rising!
Crabs are getting crabby, go figure.  The crabs apparently designed their submarines in a way that causes more fuel to be burned, the further they travel.  Traveling 1 unit costs 1 fuel, traveling 1 more unit costs 2 (total of 3). And so on and so forth.  Determine the horizontal for them to 'meet in the middle'.

In [6]:
print(f"Sample Mean: {sample_series.mean()}")
print(f"Rounded Sample Mean: {round(sample_series.mean())}")

Sample Mean: 4.9
Rounded Sample Mean: 5


In [7]:
# it appears that ths mean of the sample, (4.9)
# rounded to the nearest whole number (5)
# happens to be the be the best position. Let's see if the same holds true for the full dataset
series_mean = series.mean()
rounded_series_mean = round(series_mean)
print(f"Series Mean: {series_mean}")
print(f"Rounded Series Mean: {rounded_series_mean}")

Series Mean: 447.5
Rounded Series Mean: 448


We will try both 447 and 448 as the mean is half way betwen them. We will start with 448 as that is what the value rounds to using the round() function.

In [8]:
df = DataFrame(series)
df.columns = ['start_pos']
df['distance'] = abs(rounded_series_mean - df.start_pos)
df['fuel_cost'] = [calc_fuel_cost(x) for x in df['distance']]
total_fuel_cost = df['fuel_cost'].sum()
print(f"Answer: {int(total_fuel_cost)}")

Answer: 88612611


This was not correct. We will try with 447

In [9]:
rounded_series_mean = 447
df = DataFrame(series)
df.columns = ['start_pos']
df['distance'] = abs(rounded_series_mean - df.start_pos)
df['fuel_cost'] = [calc_fuel_cost(x) for x in df['distance']]
total_fuel_cost = df['fuel_cost'].sum()
print(f"Answer: {int(total_fuel_cost)}")

Answer: 88612508


Huzzah! We have found the correct answer!